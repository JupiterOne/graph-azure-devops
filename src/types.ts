import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface ADOIntegrationConfig extends IntegrationInstanceConfig {
  /**
   * https://dev.azure.com/ + an Azure Boards organization name (ex: "https://dev.azure.com/jupiterone")
   */
  orgUrl: string;

  /**
   * The personal access token used to authenticate requests.
   * Can be generated at {{`orgUrl`}}/_usersSettings/tokens.
   * Requires `Read` permissions for `Work Items` and `Project and Team`
   */
  accessToken: string;
  alertSeverities?: string;
}
