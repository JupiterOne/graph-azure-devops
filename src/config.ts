import { IntegrationIngestionConfigFieldMap } from '@jupiterone/integration-sdk-core';
import { INGESTION_SOURCE_IDS } from './constants';

export const ingestionConfig: IntegrationIngestionConfigFieldMap = {
  [INGESTION_SOURCE_IDS.BUILD]: {
    title: 'Build settings',
    description: 'Gather information related to the projects build settings.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.ALERTS]: {
    title: 'Devops alerts',
    description: 'Gather information about repository alerts.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.SERVICE]: {
    title: 'Devops service',
    description:
      'Creates structural relationships between repositories and projects.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.ENVIRONMENTS]: {
    title: 'Environments',
    description: 'Ingest devops project environments.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.PIPELINES]: {
    title: 'Pipelines',
    description: 'Ingest devops project pipelines.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.PROJECTS]: {
    title: 'Projects',
    description: 'Ingest devops projects.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.PULL_REQUESTS]: {
    title: 'Pull Requests',
    description: 'Ingest pull requests and information about their state.',
    defaultsToDisabled: true,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.REPOS]: {
    title: 'Repositories',
    description: 'Ingest azure devops repositories.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.TEAMS]: {
    title: 'Teams',
    description: 'Ingest teams and links users that belong to them.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.USERS]: {
    title: 'Users',
    description: 'Ingest azure devops users.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
  [INGESTION_SOURCE_IDS.WORK_ITEMS]: {
    title: 'Work items',
    description: 'Ingest work items.',
    defaultsToDisabled: false,
    cannotBeDisabled: false,
  },
};
