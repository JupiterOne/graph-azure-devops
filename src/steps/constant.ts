import {
  RelationshipClass,
  RelationshipDirection,
  StepEntityMetadata,
  StepMappedRelationshipMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const PullRequestStatus = {
  /**
   * Status not set. Default state.
   */
  0: 'NotSet',
  /**
   * Pull request is active.
   */
  1: 'Active',
  /**
   * Pull request is abandoned.
   */
  2: 'Abandoned',
  /**
   * Pull request is completed.
   */
  3: 'Completed',
  /**
   * Used in pull request search criteria to include all statuses.
   */
  4: 'All',
};

export const AlertType = {
  /**
   * The code has an unspecified vulnerability type
   */
  0: 'Unknown',
  /**
   * The code uses a dependency with a known vulnerability.
   */
  1: 'Dependency',
  /**
   * The code contains a secret that has now been compromised and must be revoked.
   */
  2: 'Secret',
  /**
   * The code contains a weakness determined by static analysis.
   */
  3: 'Code',
};

export const AlertState = {
  /**
   * Alert is in an indeterminate state
   */
  0: 'Unknown',
  /**
   * Alert has been detected in the code
   */
  1: 'Active',
  /**
   * Alert was dismissed by a user
   */
  2: 'Dismissed',
  /**
   * The issue is no longer detected in the code
   */
  4: 'Fixed',
  /**
   * The tool has determined that the issue is no longer a risk
   */
  8: 'AutoDismissed',
};

export const AlertSeverity = {
  0: 'Low',
  1: 'Medium',
  2: 'High',
  3: 'Critical',
  4: 'Note',
  5: 'Warning',
  6: 'Error',
};

export const Steps = {
  FETCH_SERVICE: 'fetch-devops-service',
  BUILD_SERVICE_ACCOUNT_RELATIONSHIP:
    'build-devOps-account-service-relationship',
  FETCH_PROJECTS: 'fetch-projects',
  FETCH_ACCOUNT: 'fetch-accounts',
  FETCH_BUILD_SETTINGS: 'fetch-build-settings',
  BUILD_PROJECT_BUILD_SETTING_RELATIONSHIP:
    'build-project-build-setting-relationship',
  FETCH_PIPELINE: 'fetch-pipelines',
  BUILD_PROJECT_PIPELINE_RELATIONSHIP: 'build-project-pipeline-relationship',
  FETCH_PULL_REQUEST: 'fetch-pullrequests',
  FETCH_REPOSITORY: 'fetch-repositories',
  BUILD_REPO_PR_RELATIONSHIP: 'build-repo-pr-relationship',
  BUILD_ACCONT_REPO_RELATIONSHIP: 'build-account-repo-relationship',
  BUILD_USER_OPENED_PR_RELATIONSHIP: 'build-user-opened-pr-relationship',
  FETCH_USERS: 'fetch-users',
  BUILD_USER_REVIEWED_PR_RELATIONSHIP: 'build-user-reviewd-pr-relationship',
  BUILD_USER_APPROVED_PR_RELATIONSHIP: 'build-user-approved-pr-relationship',
  FETCH_ALERTS: 'fetch-alerts',
  BUILD_REPO_ALERT_RELATIONSHIP: 'build-repo-alert-relationship',
};

export const Entities: Record<
  | 'DEVOPS_SERVICE_ENTITY'
  | 'PROJECT_ENTITY'
  | 'ACCOUNT_ENTITY'
  | 'BUILD_SETTING_ENTITY'
  | 'PIPELINE_ENTITY'
  | 'PULL_REQUEST_ENTITY'
  | 'REPOSITORY_ENTITY'
  | 'USER_ENTITY'
  | 'ALERT_ENTITY',
  StepEntityMetadata
> = {
  DEVOPS_SERVICE_ENTITY: {
    resourceName: 'AzureDevOps',
    _type: 'azure_devops',
    _class: ['Service'],
  },
  PROJECT_ENTITY: {
    resourceName: 'ADO Project',
    _type: 'azure_devops_project',
    _class: 'Project',
  },
  ACCOUNT_ENTITY: {
    resourceName: 'Azure Devops Account',
    _type: 'azure_devops_account',
    _class: 'Account',
  },
  BUILD_SETTING_ENTITY: {
    resourceName: 'AzureBuildSettings',
    _type: 'azure_devops_build_settings',
    _class: 'Configuration',
  },
  PIPELINE_ENTITY: {
    resourceName: 'AzureDevOpsPipeline',
    _type: 'azure_devops_pipeline',
    _class: 'Workflow',
  },
  PULL_REQUEST_ENTITY: {
    resourceName: 'AzureDevOpsPullRequest',
    _type: 'azure_devops_pr',
    _class: 'PR',
  },
  REPOSITORY_ENTITY: {
    resourceName: 'Repository',
    _type: 'azure_devops_repo',
    _class: 'Repository',
  },
  USER_ENTITY: {
    resourceName: 'ADO User',
    _type: 'azure_devops_user',
    _class: 'User',
  },
  ALERT_ENTITY: {
    resourceName: 'Azure DevOps Alerts',
    _type: 'azure_devops_alert_finding',
    _class: 'Finding',
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_USERS'
  | 'USERS_MANAGES_ACCOUNT'
  | 'AZURE_DEVOPS_ACCOUNT_HAS_DEVOPS_SERVICE'
  | 'AZURE_DEVOPS_SCANS_PROJECTS'
  | 'AZURE_DEVOPS_ACCOUNT_HAS_PROJECTS'
  | 'PROJECT_HAS_BUILD_SETTINGS'
  | 'AZURE_DEVOPS_PROJECT_HAS_PIPELINE'
  | 'REPO_HAS_PULL_REQUEST'
  | 'PROJECT_HAS_REPO'
  | 'ACCOUNT_OWNS_REPO'
  | 'USER_OPENED_PR'
  | 'USER_REVIEWED_PR'
  | 'USER_APPROVED_PR'
  | 'REPO_HAS_ALERTS',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_USERS: {
    _type: 'azure_devops_account_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.USER_ENTITY._type,
  },
  USERS_MANAGES_ACCOUNT: {
    _type: 'azure_devops_user_manages_account',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.ACCOUNT_ENTITY._type,
  },
  AZURE_DEVOPS_ACCOUNT_HAS_DEVOPS_SERVICE: {
    _type: 'azure_devops_account_has_devops',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.DEVOPS_SERVICE_ENTITY._type,
  },
  AZURE_DEVOPS_SCANS_PROJECTS: {
    _type: 'azure_devops_scans_project',
    _class: RelationshipClass.SCANS,
    sourceType: Entities.DEVOPS_SERVICE_ENTITY._type,
    targetType: Entities.PROJECT_ENTITY._type,
  },
  AZURE_DEVOPS_ACCOUNT_HAS_PROJECTS: {
    _type: 'azure_devops_account_has_project',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.PROJECT_ENTITY._type,
  },
  PROJECT_HAS_BUILD_SETTINGS: {
    _type: 'azure_devops_project_has_build_settings',
    _class: RelationshipClass.HAS,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: Entities.BUILD_SETTING_ENTITY._type,
  },
  AZURE_DEVOPS_PROJECT_HAS_PIPELINE: {
    _type: 'azure_devops_project_has_pipeline',
    _class: RelationshipClass.HAS,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: Entities.PIPELINE_ENTITY._type,
  },
  REPO_HAS_PULL_REQUEST: {
    _type: 'azure_devops_repo_has_pr',
    _class: RelationshipClass.HAS,
    sourceType: Entities.REPOSITORY_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  PROJECT_HAS_REPO: {
    _type: 'azure_devops_project_uses_repo',
    _class: RelationshipClass.USES,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: Entities.REPOSITORY_ENTITY._type,
  },
  ACCOUNT_OWNS_REPO: {
    _type: 'azure_devops_account_owns_repo',
    _class: RelationshipClass.OWNS,
    sourceType: Entities.ACCOUNT_ENTITY._type,
    targetType: Entities.REPOSITORY_ENTITY._type,
  },
  USER_OPENED_PR: {
    _type: 'azure_devops_user_opened_pr',
    _class: RelationshipClass.OPENED,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  USER_REVIEWED_PR: {
    _type: 'azure_devops_user_reviewed_pr',
    _class: RelationshipClass.REVIEWED,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  USER_APPROVED_PR: {
    _type: 'azure_devops_user_approved_pr',
    _class: RelationshipClass.APPROVED,
    sourceType: Entities.USER_ENTITY._type,
    targetType: Entities.PULL_REQUEST_ENTITY._type,
  },
  REPO_HAS_ALERTS: {
    _type: 'azure_devops_repo_has_alert_finding',
    _class: RelationshipClass.HAS,
    sourceType: Entities.REPOSITORY_ENTITY._type,
    targetType: Entities.ALERT_ENTITY._type,
  },
};

export const MappedRelationships: Record<
  'PROJECT_USES_GITHUB_REPO' | 'PROJECT_USES_BITBUCKET_REPO',
  StepMappedRelationshipMetadata
> = {
  PROJECT_USES_GITHUB_REPO: {
    _type: 'azure_devops_project_uses_repo',
    _class: RelationshipClass.USES,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: 'github_repo',
    direction: RelationshipDirection.FORWARD,
  },
  PROJECT_USES_BITBUCKET_REPO: {
    _type: 'azure_devops_project_uses_repo',
    _class: RelationshipClass.USES,
    sourceType: Entities.PROJECT_ENTITY._type,
    targetType: 'bitbucket_repo',
    direction: RelationshipDirection.FORWARD,
  },
};
