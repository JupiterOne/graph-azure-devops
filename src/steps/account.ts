import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { ADOIntegrationConfig } from '../types';
import { Entities, Steps } from './constant';

export function getAccountKey(uniqueId) {
  return `azure-devops-account:${uniqueId}`;
}

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<ADOIntegrationConfig>) {
  const name = `Azure Devops - ${instance.name}`;
  const accountEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: 'azure-devops',
          name: 'Azure Devops Account',
        },
        assign: {
          _key: getAccountKey(instance.id),
          _type: Entities.ACCOUNT_ENTITY._type,
          _class: 'Account',
          name,
          displayName: name,
        },
      },
    }),
  );
  await jobState.setData(Entities.ACCOUNT_ENTITY._type, accountEntity);
}

export const accountSteps: IntegrationStep<ADOIntegrationConfig>[] = [
  {
    id: Steps.FETCH_ACCOUNT,
    name: 'Fetch Account Details',
    entities: [Entities.ACCOUNT_ENTITY],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
