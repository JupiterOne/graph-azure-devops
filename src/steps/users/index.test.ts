import {
  createMockStepExecutionContext,
  MockIntegrationStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { fetchUsers } from '.';
import { config } from '../../../test/config';
import { setupAzureRecording } from '../../../test/recording';
import { getMatchRequestsBy } from '../../../test/getMatchRequestsBy';
import { testEntities } from '../../../test/testEntities';
import { testRelationships } from '../../../test/testRelationships';
import { toArray } from '../../../test/toArray';
import { ADOIntegrationConfig } from '../../types';
import { fetchAccountDetails } from '../account';
import { Entities } from '../constant';

let recording: Recording;

afterEach(async () => {
  if (recording) {
    await recording.stop();
  }
});

describe('fetchUsers', () => {
  async function setupFetchUsersContext(): Promise<
    MockIntegrationStepExecutionContext<ADOIntegrationConfig>
  > {
    const setupContext = createMockStepExecutionContext({
      instanceConfig: config,
    });

    await fetchAccountDetails(setupContext);

    const accountEntity = setupContext.jobState.collectedEntities.filter((e) =>
      toArray(e._class).includes('Account'),
    )[0];

    return createMockStepExecutionContext({
      instanceConfig: config,
      entities: setupContext.jobState.collectedEntities,
      relationships: setupContext.jobState.collectedRelationships,
      setData: {
        [Entities.ACCOUNT_ENTITY._type]: accountEntity,
      },
    });
  }

  it('Should create user entities and relationships', async () => {
    recording = setupAzureRecording({
      directory: __dirname,
      name: 'fetchUsers',
      options: {
        matchRequestsBy: getMatchRequestsBy({ config: config }),
      },
    });

    const context = await setupFetchUsersContext();

    await fetchUsers(context);

    const userEntities = context.jobState.collectedEntities;
    testEntities(userEntities, 'userEntities');
    const userRelationships = context.jobState.collectedRelationships;
    testRelationships(userRelationships, 'userRelationships');
  });
});
