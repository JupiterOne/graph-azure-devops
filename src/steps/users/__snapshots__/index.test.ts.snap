// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchUsers Should create user entities and relationships: userEntities 1`] = `
[
  {
    "_class": [
      "User",
    ],
    "_key": "e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
    "_rawData": [
      {
        "name": "default",
        "rawData": {
          "identity": {
            "_links": {
              "avatar": {
                "href": "https://dev.azure.com/michaelknoedel/_apis/GraphProfile/MemberAvatars/aad.Njg3NTczODAtOTcxMi03YWU4LWFmNmItNGRmZWExZWM4Y2Fj",
              },
            },
            "descriptor": "aad.Njg3NTczODAtOTcxMi03YWU4LWFmNmItNGRmZWExZWM4Y2Fj",
            "displayName": "Michael Knoedel",
            "id": "e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
            "imageUrl": "https://dev.azure.com/michaelknoedel/_api/_common/identityImage?id=e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
            "uniqueName": "michael.knoedel@jupiterone.com",
            "url": "https://spsprodcus5.vssps.visualstudio.com/Ae6472b42-435c-4da8-bb39-ae8a56ec4bbd/_apis/Identities/e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
          },
        },
      },
    ],
    "_type": "azure_devops_user",
    "active": true,
    "createdOn": undefined,
    "descriptor": "aad.Njg3NTczODAtOTcxMi03YWU4LWFmNmItNGRmZWExZWM4Y2Fj",
    "displayName": "Michael Knoedel",
    "email": "michael.knoedel@jupiterone.com",
    "emailDomain": [
      "jupiterone.com",
    ],
    "imageLink": "https://dev.azure.com/michaelknoedel/_api/_common/identityImage?id=e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
    "name": "Michael Knoedel",
    "profileLink": undefined,
    "username": "michael.knoedel@jupiterone.com",
    "webLink": "https://spsprodcus5.vssps.visualstudio.com/Ae6472b42-435c-4da8-bb39-ae8a56ec4bbd/_apis/Identities/e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
  },
]
`;

exports[`fetchUsers Should create user entities and relationships: userRelationships 1`] = `
[
  {
    "_class": "HAS",
    "_fromEntityKey": "azure-devops-account:local-integration-instance",
    "_key": "azure-devops-account:local-integration-instance|has|e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
    "_toEntityKey": "e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
    "_type": "azure_devops_account_has_user",
    "displayName": "HAS",
  },
  {
    "_class": "MANAGES",
    "_fromEntityKey": "e7f84ec3-0ecc-4b78-878b-42a5c2c8f424",
    "_key": "e7f84ec3-0ecc-4b78-878b-42a5c2c8f424|manages|azure-devops-account:local-integration-instance",
    "_toEntityKey": "azure-devops-account:local-integration-instance",
    "_type": "azure_devops_user_manages_account",
    "displayName": "MANAGES",
  },
]
`;
